lenskit-core/src/main/java/org/lenskit/util/Text.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2016 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Utf8;
+import org.joda.convert.FromString;
+import org.joda.convert.ToString;
+
+import java.io.Serializable;
+import java.text.Normalizer;
+import java.util.Arrays;
+
+/**
+ * UTF-8 encoded text string.
+ */
+public final class Text implements Serializable {
```

Comment:
Yes, we did. I found 2 other implementations of UTF-8 strings for Java: one lives in Hadoop, and the other lives in Apache Avro, which is a full serialization framework that we don't otherwise need.

So I built it, and was none too pleased about the need to do so.

But it will significantly reduce our memory use when storing a bunch of strings in entities in memory, because Java uses UTF-16.